spec/integration/login_spec.rb

```diff
@@ -0,0 +1,141 @@
+
+    # Pulling images that we should already have (so we go faster :P). Then
+    # re-tag it so they can be pushed. We tag both a pusheable image and
+    # another one that cannot be pushed (namespace does not exist).
+    imagename   = "registry"
+    pulled_tags = ["2.2.1", "2.3"]
+    pulled_tags.each do |tag|
+      base  = "#{imagename}:#{tag}"
+      img   = "library/#{base}"
+
+      pull(img)
+      system("docker tag #{img} #{registry_hostname}/#{img}")
+      system("docker tag #{img} #{registry_hostname}/#{base}")
+    end
+
+    # Pushing...
+    expect(push("#{registry_hostname}/#{imagename}")).to be_truthy
+    expect(push("#{registry_hostname}/library/#{imagename}", false)).to be_truthy
+
+    # And finally let's see what's inside the DB. The registry might not be
+    # ready yet, so let's call it inside of `eventually_expect`.
+    repos = []
+    tags = []
+    eventually_expect(1) do
+      repos = rails_exec("Repository.all.to_json")
+      repos.size
+    end
+    eventually_expect(2) do
+      tags = rails_exec("Tag.all.to_json")
+      tags.size
+    end
+
+    expect(repos.first["name"]).to eq "registry"
+    names = tags.map { |t| t["name"] }
+    expect(names).to eq pulled_tags
+  end
+
+  it "handles the personal namespace properly" do
+    create_user(name, email, password, true)
+    expect { login(name, password, email) }.not_to raise_error
+
+    # Push an image to the personal namespace.
+    img     = "registry:2.3"
+    target  = "#{registry_hostname}/#{name}/#{img}"
+    pull(img)
+    system("docker tag #{img} #{target}")
+    expect(push(target)).to be_truthy
+
+    # Logout and try to pull/push it. It should fail.
+    logout!
+    expect(spawn_cmd("docker pull #{target}")).to be_falsey
+    expect(push(target, false)).to be_truthy
+  end
+
+  it "handles viewers, contributors and owners accordingly" do
+    # TODO: once create_user works on LDAP, remove this guard.
+    unless ldap?
+      create_user(name, email, password, true)
+      create_user(name + "1", "test1@email.com", password)
+      create_user(name + "2", "test2@email.com", password)
+
+      rb = <<HERE
+team = Team.new(name: 'team')
+user1 = User.find_by(username: '#{name}')
+user2 = User.find_by(username: '#{name}1')
+user3 = User.find_by(username: '#{name}2')
+team.owners = [user1]
+team.contributors = [user2]
+team.viewers = [user3]
+team.save
+Namespace.create(team: team.reload, name: 'namespace', registry: Registry.get)
+HERE
+      docker_exec("integration_portus", "rails r \"#{rb}\"")
+
+      expect { login(name, password, email) }.not_to raise_error
+
+      # Push an image to the team namespace.
+      img     = "registry:2.3"
+      target  = "#{registry_hostname}/namespace/#{img}"
+      pull(img)
+      system("docker tag #{img} #{target}")
+
+      # Owner can push.
+      expect(push(target)).to be_truthy
+      logout!
+
+      # Contributor can push.
+      expect { login(name + "1", password, email) }.not_to raise_error
```

Comment:
I found a bit confusing the way variables are named, especially to remember `user` is owner, `user` is contributor,...

Why don't we change `create_user` to return an hash with `:name`, `:password` and `:email` and we store that return value into variables named `owner`, `contributor`,...